packages/ui-scripts/src/publish.ts

```diff
@@ -62,73 +65,140 @@ async function publish({
-    try {
-      info(`📦  Version: ${version}, Tag: snapshot`)
-      return await publishPackages(packageName, 'prerelease', 'snapshot')
-    } catch (e: any) {
-      error(e)
-      process.exit(1)
+    info(`📦  Version: ${version}, Tag: snapshot`)
+    return publishSnapshotVersion({
+      version,
+      packageName,
+      packages
+    })
+  }
+}
+
+/**
+ * Publishes each package to npm.
+ */
+async function publishRegularVersion(arg: {
+  version: string
+  tag: string
+  packages: PackageInfo[]
+}) {
+  const { version, packages, tag } = arg
+  for await (const pkg of publishPackages(packages, version, tag)) {
+    info(`📦  Version ${version} of ${pkg.name} was successfully published!`)
+  }
+}
+
+/**
+ * Calculates the new snapshot version based on the latest tag
+ * and the current commit, then publishes each package to npm
+ * with the new snapshot version.
+ */
+async function publishSnapshotVersion(arg: {
+  version: string
+  packageName: string
+  packages: PackageInfo[]
+}) {
+  const { version, packageName, packages } = arg
+  const snapshotVersion = calculateNextSnapshotVersion(version)
+
+  info(`applying new snapshot version (${snapshotVersion}) to each package`)
+
+  await bumpPackages(packageName, snapshotVersion)
+
+  for await (const pkg of publishPackages(packages, snapshotVersion)) {
+    info(
+      `📦  Version ${snapshotVersion} of ${pkg.name} was successfully published!`
+    )
+  }
+}
+
+/**
+ * Calculates the new snapshot version.
+ * @returns the new snapshot version
+ */
+function calculateNextSnapshotVersion(version: string): string {
+  const ver = `v${version}`
+  // get the commit count between the current released version
+  // and the commit that we are on currently
+  // note: this will not include the release commit into the count
+  const commitCount = runGitCommand([
+    'rev-list',
+    '--count',
+    `${ver}..${runGitCommand(['describe'])}`
+  ])
+
+  // we have to decrease the commitCount by 1 because the snapshots are
+  // zero based
+  const snapshotVersion = `${semver.inc(version, 'patch')}-snapshot-${
+    Number(commitCount) - 1
+  }`
+
+  info(`next snapshot version is: ${snapshotVersion}`)
+
+  return snapshotVersion
+}
+
+/**
+ * An async generator function which will do the publishing
+ * for each pacakge.
+ */
+async function* publishPackages(
+  packages: PackageInfo[],
+  version: string,
+  tag?: string
+) {
+  const wait = (delay: number) =>
+    new Promise((resolve) => {
+      setTimeout(resolve, delay)
+    })
+
+  for (const pkg of packages) {
+    const { stdout } = await runCommandAsync(
+      'npm',
+      ['info', pkg.name, '--json'],
+      [],
+      {
+        stdio: 'pipe'
+      }
+    )
+    const packageVersions: string[] = JSON.parse(stdout).versions
+
+    // if the package is already in the registry then don't try
+    // to publish it again
+    if (packageVersions.includes(version)) {
+      info(`📦  v${version} of ${pkg.name} is already published!`)
```

Comment:
hmm we should discuss how we handle this case. I think last time the release failed because for some reason empty packages were released. I'm leaning towards aborting with a huge error message in this case, normally we should never try to publish over a package.

Reply:
The last version had a bug in it, so it tried to release a version that already existed in the registry.
This new version will not even try to release the package that is in the registry.

Reply:
As discussed in the daily, I added the code to throw an error when this check is true.